### PR TITLE
Add config flag to disable usage of endpointslices

### PR DESCRIPTION
**Pull request by Christoph-Raab**
As discussed in https://github.com/metallb/metallb/issues/936 there are certain scenarios where endpointslices are technical available but should not be used by metallb.

This PR adds a config flag to disable the usage of endpointslices completely. If the flag is not provided the behaviour is the same as before.

### CODE REVIEW

**Comment by russellb** [comment]
@fedepaol Can you take a look?

**Comment by fedepaol** [comment]
> @fedepaol Can you take a look?

sure!

**Comment by fedepaol** [comment]
@Christoph-Raab if I get this right, in the cluster you are mentioning the EndpointSliceProxying feature gate is disabled but EndpointSlice is enabled, so the CR is getting filled but it's not used.

Just to understand, any reason why you did not go with disabling EndpointSlice too instead of modifying metallb?

**Comment by Christoph-Raab** [comment]
> Just to understand, any reason why you did not go with disabling EndpointSlice too instead of modifying metallb?

The main reason is that you can not disable the Endpointslices resource in the apiserver. So we would not only need to disable the endpointslices controller in the controller-manager, but also deny creating the resource and remove all existing ones. 

That's why we want a feature flag like core-dns has to disable the usage in the addon.

**Comment by fedepaol** [comment]
> > if I get this right, in the cluster you are mentioning the EndpointSliceProxying feature gate is disabled but EndpointSlice is enabled, so the CR is getting filled but it's not used.
> 
> Exactly.
> 
> > Just to understand, any reason why you did not go with disabling EndpointSlice too instead of modifying metallb?
> 
> The main reason is that you can not disable the Endpointslices resource in the apiserver. So we would not only need to disable the endpointslices controller in the controller-manager, but also deny creating the resource and remove all existing ones.
> 
> That's why we want a feature flag like core-dns has to disable the usage in the addon.

But (and I may be wrong here), if you disable the controller in controller-manager via the EndpointSlice, ep slices won't be created (even though the resource is still available), and the automatic check should cover that:

https://github.com/metallb/metallb/blob/4de2c0c7fbd3e4eef624bdff40c1c02f98d252d8/internal/k8s/k8s.go#L611

Here we check if the kubernetes service doesn't have endpoint slices, and if so we assume that ep slices are disabled. Would that work or is that logic bugged somehow?

**Comment by Christoph-Raab** [comment]
You are correct that if you disable the controller, EndpointSlices will not be automatically created by kubernetes. 

But the resource still exists in the apiserver. So any user or addon/operator could still create EndpointSlices. So you need to forbid that operation as well, remove all existing ones (because they are no longer manged by the controller/user/addon/operator) and hope, that no user/addon/operator depends on them or something breaks, if they are not there/can not be created.

It's much saver to disable them in metallb, where we are certain, that we don't want to use them and nothing breaks if we do so.

**Comment by fedepaol** [comment]
> You are correct that if you disable the controller, EndpointSlices will not be automatically created by kubernetes.
> 
> But the resource still exists in the apiserver. So any user or addon/operator could still create EndpointSlices. So you need to forbid that operation as well, remove all existing ones (because they are no longer manged by the controller/user/addon/operator) and hope, that no user/addon/operator depends on them or something breaks, if they are not there/can not be created.
>

Well, endpoints slices are not meant to be created by anything else than k8s. You could say the same thing about endpoints, if somebody creates a new endpoint manually you break the service.
In particular, the check we are performing is against the "kubernetes" service which is always there.

I understand on the other hand the fact that having to delete them is clunky, and that we did not think about this corner case where a cluster could switch from having them enabled to disabled.
 
> It's much saver to disable them in metallb, where we are certain, that we don't want to use them and nothing breaks if we do so.

File: controller/main.go [hunk 1]
```diff
@@ -122,13 +122,14 @@ func (c *controller) MarkSynced(l log.Logger) {
...
-		deployName = flag.String("deployment", os.Getenv("METALLB_DEPLOYMENT"), "name of the MetalLB controller Deployment")
-		logLevel   = flag.String("log-level", "info", fmt.Sprintf("log level. must be one of: [%s]", strings.Join(logging.Levels, ", ")))
+		port          = flag.Int("port", 7472, "HTTP listening port for Prometheus metrics")
+		config        = flag.String("config", "config", "Kubernetes ConfigMap containing MetalLB's configuration")
+		namespace     = flag.String("namespace", os.Getenv("METALLB_NAMESPACE"), "config / memberlist secret namespace")
+		kubeconfig    = flag.String("kubeconfig", "", "absolute path to the kubeconfig file (only needed when running outside of k8s)")
+		mlSecret      = flag.String("ml-secret-name", os.Getenv("METALLB_ML_SECRET_NAME"), "name of the memberlist secret to create")
+		deployName    = flag.String("deployment", os.Getenv("METALLB_DEPLOYMENT"), "name of the MetalLB controller Deployment")
+		logLevel      = flag.String("log-level", "info", fmt.Sprintf("log level. must be one of: [%s]", strings.Join(logging.Levels, ", ")))
+		disableSlices = flag.Bool("disable-slices", false, "Disable the usage of endpointslices")
```
Comment (fedepaol) [inline controller/main.go:132]: I'd call this `disable-epslices` and mention the fact that it defaults to endpoints in the description without relying on the autodiscovery mechanism

Comment (fedepaol) [reply controller/main.go:132]: Also, EndpointSlices or endpoint slices in the description

File: internal/k8s/k8s.go [hunk 2]
```diff
@@ -603,7 +604,11 @@ func serviceNameForSlice(endpointSlice *discovery.EndpointSlice) (string, error)
 }
 
 // UseEndpointSlices detect if Endpoints Slices are enabled in the cluster.
-func UseEndpointSlices(kubeClient kubernetes.Interface) bool {
+func UseEndpointSlices(kubeClient kubernetes.Interface, cfg *Config) bool {
+	if cfg.DisableSlices {
```
Comment (fedepaol) [inline internal/k8s/k8s.go:608]: Can you add a comment about the particular scenario (epslices enabled but disabled in the proxy) that we are covering with this parameter?

File: controller/main.go (hunk 1 above)
Comment (Christoph-Raab) [reply controller/main.go]: Done

File: internal/k8s/k8s.go (hunk 2 above)
Comment (Christoph-Raab) [reply internal/k8s/k8s.go]: Done

File: internal/k8s/k8s.go [hunk 3]
```diff
@@ -164,7 +165,7 @@ func New(cfg *Config) (*Client, error) {
 		c.syncFuncs = append(c.syncFuncs, c.svcInformer.HasSynced)
 
 		if cfg.ReadEndpoints {
-			if !UseEndpointSlices(c.client) {
+			if !UseEndpointSlices(c.client, cfg) {
```
Comment (fedepaol) [inline internal/k8s/k8s.go:168]: I'd change this to check the flag outside, seems more readable.

Something like `if cfg.DisableEPSlices || !UseEndpointSlices(c.client) {`

**Comment by fedepaol** [comment]
@Christoph-Raab sorry for not reviewing this before

**Comment by fedepaol** [comment]
> @Christoph-Raab sorry for not reviewing this before

Also, mind adding the rationale to the commit message? Same thing you added as a comment.

File: internal/k8s/k8s.go (hunk 3 above)
Comment (Christoph-Raab) [reply internal/k8s/k8s.go:168]: Having all logic to determine if EndpointSlices should be used in one method seams like the reasonalbe thing to do, so I put it there. But I can change that.

Should I add a comment like I did in `UseEndpointSlices`? Or is it fine without that?

Comment (fedepaol) [reply internal/k8s/k8s.go:168]: Agree, but passing a flag just to return early won't add much value to the function itself I think. Might be different if we start using that in many different places, but that's not the current situation.

I'd move the comment too.

Comment (Christoph-Raab) [reply internal/k8s/k8s.go]: Done

**Comment by Christoph-Raab** [comment]
> > @Christoph-Raab sorry for not reviewing this before
> 
> Also, mind adding the rationale to the commit message? Same thing you added as a comment.

Sure, no problem

**Comment by russellb** [comment]
just to be explicit, i'm looking for a "lgtm" from @fedepaol since @fedepaol did the endpointslices support

**Comment by fedepaol** [comment]
lgtm , and thanks!

**Review by rata — APPROVED** [review]
LGTM, thanks @Christoph-Raab and @fedepaol ! Will merge now.

However, some questions for both regarding a possible follow-up PR:
1. Would it make sense in a future PR to try to detect this situation in "UseEndpointSlices()" without any need for a user-specified flag?

For example, we can look at the kubernetes service (to say an example) or some other service that should be present on any conformant clusters and see if endpoint slices are created for them? Or was that considered too hacky or there isn't such a service we can know to be present?


Or the issue is that endpoint slices are being created and updated, but we don't want to use them in metallb for some reason? What is the reason to not use them in metalLB if they are created and updated by k8s controllers?

I think I'm missing those details. Will merge this now, but please let me know what you think :)

**Comment by rata** [comment]
Merging, LGTM to me, also LGTM from @fedepaol and IIUC also @russellb

**Comment by fedepaol** [comment]
> For example, we can look at the kubernetes service (to say an example) or some other service that should be present on any conformant clusters and see if endpoint slices are created for them? Or was that considered too hacky or there isn't such a service we can know to be present?
> 
> Or the issue is that endpoint slices are being created and updated, but we don't want to use them in metallb for some reason? What is the reason to not use them in metalLB if they are created and updated by k8s controllers?
> 

The current code tries to do what you just described, against the k8s service, checking for the presence of epslices.

There is one corner case though, where you can have the epslices enabled but kubeproxy instructed to consume endpoints, that is enabled by a feature flag and there's no way (I know of) to find it programmatically.
I asked @Christoph-Raab to specify that in the commit message and as a comment in the code.

> I think I'm missing those details. Will merge this now, but please let me know what you think :)

**Comment by rata** [comment]
@fedepaol oh, you are right, on my quick look I missed that the current code is doing just that. Thanks!

@Christoph-Raab just curious, in which scenario endpoint slices are created by the k8s controllers (I assume they have up to date information) but is problematic if MetalLB uses them? Do you hit a bug in MetalLB using endpoint slices? Or why MetalLB should not use them?

In any case, if kube-proxy is doing this, I guess we can too...

**Comment by champtar** [comment]
(on my phone) Is the flag name the same as kube-proxy ? If not does it make sense to use the same flag name ?

**Comment by Christoph-Raab** [comment]
@rata 

> @Christoph-Raab just curious, in which scenario endpoint slices are created by the k8s controllers (I assume they have up to date information) but is problematic if MetalLB uses them? Do you hit a bug in MetalLB using endpoint slices? Or why MetalLB should not use them?

We disabled endpointslices with kube-proxy but not in the whole cluster. So metallb would be using them but kube-proxy would still work with endpoints. We don't really trust that endpoints and endpointslices are in perfect sync so we rather want addons to use endpoints.

@champtar 

> Is the flag name the same as kube-proxy ? If not does it make sense to use the same flag name ?

Kube-proxy uses a flag in the config map:

```
    featureGates:
      EndpointSliceProxying: false
```

We chose `--disable-epslices` for metallb which describes quite good what it does.

**Pull request merged**